### PR TITLE
[WIP] Some thoughts about a FunctionalAI

### PR DESCRIPTION
**Pull request by Naios**
# Introduction

In the last weeks i thought about improving c++ scripting in general.
My idea was to bring some amazing new c++11 features to ai scripting which are missing, especially `std::function` (lamdas!), `std::bind` and `std::chrono`, mixed with some metaprogramming.

The proposed improvements don't depend on each other, we could merge it seperate.
# Improving the EventMap
## Comparing the current solutions

In this time the `EventMap` fulfill its role to inform about events that needs to be executed.
Everybody who went into CreatureAI scripting will know that it lacks some kind of useability to define events:
- First define an enum value `EventMap` can't handle 0 events, every event needs to be >= 1.
- Every event needs an entry in the corresponding switch clause.
- Time is always in ms, in most cases its not very readable.
- Lacks useablity to repeat events properly inside the switch clause (it never worked for me personally). `EventMap::Repeat` is used only 10 times and only in one script (`boss_kelthuzad.cpp`). `EventMap::ScheduleEvent` is used in most cases instead.

Another approach to solve scheduling is the `Unit::EventProcessor` (if you don't know it yet check out the [implementation of Algalon](https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/scripts/Northrend/Ulduar/Ulduar/boss_algalon_the_observer.cpp#L480)).
It works by passing functional objects to the `EventProcessor` which will execute it after the time has passed. I liked it much and used it several times as well.
But it has some disadvantages:
- You have to define a functional object every time you are using a new event.
- The `EventProcessor` lacks the ability to reschedule events or to cancel it.
## Introducing the Function Scheduler

This several disadvantages of the current solutions lead me to invent a scheduler which uses c++11's `std::function` (**lamdas** or **`std::bind`**) to execute the actual event and `std::chrono` (new in c++11 as well) to pass the time duration.
It targets useability and convenience.

This will make scheduling of events very easy:

``` c++
FunctionScheduler scheduler;

void EnterCombat(Unit* /*who*/) override
{
    scheduler.Schedule(Seconds(5), [this](TaskContext context)
    {
        me->Yell("Hey, this is a simple function scheduler test", LANG_UNIVERSAL);
    });
}

void UpdateAI(uint32 diff) override
{
    // It can be used in every ai.
    scheduler.Update(Milliseconds(diff));
}
```

It supports named and anonymous events.
We can also assign an event name (groupid) to the function:

``` c++
enum MyEvents
{
    EVENT_DO_SOMETHING_0, // 0 As group (event id) is supported (internally uses an optional class).
    EVENT_DO_SOMETHING_1,
};

void doSomething()
{
    // all common time durations are supported.
    scheduler.Schedule(Minutes(1), EVENT_DO_SOMETHING_0, [this](TaskContext context)
    {
        me->Yell("Hey, its 1 minute later....", LANG_UNIVERSAL);

        // Repeat it with the same duration.
        context->Repeat();
    });

    scheduler.DelayGroup(EVENT_DO_SOMETHING_0, Seconds(30));

    scheduler.CancelGroup(EVENT_DO_SOMETHING_0);

    // Or
    // scheduler.CancelAll();
}
```

As soon as the function is executed a `TaskContext` object is passed to the function which will offers the possibility to modify the event itself (repeat it with the same or a new duration) or schedule new events from within the context (events scheduled from a context use the timepoint where the task was supposed to be executed and not the current time -> correct scheduling even the world update diff is 2 seconds).

``` c++
void doSomething()
{
    scheduler.Schedule(Seconds(5), [this](TaskContext context)
    {
        me->Yell("Repeat this every 10 s", LANG_UNIVERSAL);

        // Most methods are lazy overloaded (supports concatenation of multiple expressions).
        scheduler
            .Schedule(Seconds(2), [this](TaskContext context)
            {
                me->Yell("After 2s...", LANG_UNIVERSAL);
            })
            .Schedule(Seconds(4), [this](TaskContext context)
            {
                me->Yell("After 4s...", LANG_UNIVERSAL);
            });

        // Repeat this with a constant time of 10s
        context->Repeat(Seconds(10));
    });
}
```

If you are looking for more usage info take a look at the [test ai script](https://github.com/Naios/TrinityCore/blob/functionalai/src/server/scripts/Custom/functional_ai_tests.cpp) or at the readme of my [fnc library](https://github.com/Naios/FnC-Cpp#function-scheduler).
# Event Listeners
## The hook system

The current hook system (you just overwrite virtual functions) is low level and very efficient.
But we could implement much better useable systems on top of it, for example a scripting language enviroment or just another c++ api which is easier to use, everything is backward compatible due to the hook usage.
## Event Listeners

The Event-Listener pattern is very simple: you can add and remove listeners which are informed if a certain event is invoked:
compositeshart

This gives us flexibility when writing complex scripts, we could add/remove listeners on phase change for example or depending on the difficulty of an instance (add a second listener that handles heroic events for example):

``` c++
AddListener([this](OnEnterCombatEvent& event)
{
    me->Yell("Hey, OnEnterCombatEvent was thrown!", LANG_UNIVERSAL);

    event.victim->GetTypeId();
});
```

The usage is very easy, every event has an event structure, `AddListener` is overloaded several times and detects which hook you want to use.

Currently there are following events available:
- OnEnterCombatEvent
- OnJustDiedEvent
- OnSpellHitEvent

This gives us several advantages:
- We could seperate different aspects of a script (Add several listeners for one event depending on the difficulty or some conditions).
- If we change the signature of an event we just need to change the structure, we wouldnt invalidate custom scripts anymore.
# Duplicated Code

When we write ai's or scripted behaviour we are forced with a lot of duplicated code, following changes try to eliminate this.
## Simple script loader

In near every script an ai is wrapped inside a Script (`CreatureScript` or `SpellScript` for example).

We could remove this duplicated code by wrapping an ai into `SimpleScriptLoader` (templated classes that provide the `CreatureScript` which is needed to get the ai to work).

We could also fetch the scriptname from the typeid of the struct which would make it optional to pass the scriptname to the loader.

``` c++
struct npc_simple_scheduler_test : public CreatureAI
{
    npc_simple_scheduler_test(Creature* creature) : CreatureAI(creature)
    {
    }

    void UpdateAI(uint32 diff) override
    {
    }
};

void AddSC_functional_ai_tests()
{
    new SimpleCreatureScriptLoader<npc_simple_scheduler_test>("npc_simple_scheduler_test");
}
```
# Changelog
- Added simple loaders
- Updated fnc to the latest rev which supports scheduling of asyncs and rescheduling now.
# Disclaimer

Most changes in this PR will change from time to time, this also goes for [my library i used](https://github.com/Naios/FnC-Cpp) to implement the scheduler and composite to test it without TrinityCore (i don't want to introduce it as dependency, maybe i will adapt it and remove the parts not needed).
Use only constructive criticism please if you disagree with this topic.

I invite everyone to discuss and think about the suggested improvements.

### CODE REVIEW

File: src/server/scripts/Custom/functional_ai_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        void EnterCombat(Unit* /*who*/) override
+        {
+            me->Yell("Mh who is that...", LANG_UNIVERSAL);
+
+            // Schedule a simple task that gets executed once in 5 seconds.
+            // As time unit std::chrono::duration is used which offers typedef for common used timeunits.
+            // No magic milliseconds anymore!
+            scheduler.Schedule(Seconds(5), [this](TaskContext& context)
```
Comment (DDuarte) [inline src/server/scripts/Custom/functional_ai_tests.cpp]: Couldn't this be written as `5s`? http://en.cppreference.com/w/cpp/chrono/operator%22%22s

Comment (Naios) [inline src/server/scripts/Custom/functional_ai_tests.cpp]: Probably it would't be a problem to implement it but its C++14.

Comment (DDuarte) [inline src/server/scripts/Custom/functional_ai_tests.cpp]: Oh, I didn't notice :/

Comment (Naios) [inline src/server/scripts/Custom/functional_ai_tests.cpp]: But i could add some special c++11 chrono legacy literals for you xD http://pastebin.com/8nx4jZem

File: sql/custom/world/.gitignore [hunk 2]
```diff
@@ -1 +0,0 @@
-*.sql
```
Comment (Shauren) [inline sql/custom/world/.gitignore]: uwot?

**Comment by Aokromes** [comment]
Failed — The Travis CI build failed

**Comment by Lordron** [comment]
I want performance comparison between this system and EventMap)

**Comment by Rochet2** [comment]
> Its possible that the new event is executed immediately!

What does this mean for events with 0 time? Can they be executed instantly?
Instant meaning right after scheduling and not waiting for the next update.
I think there are scripts that use the ability to schedule an event for the "next tick" with using 0 delay.
Regex to find some: `ScheduleEvent\([^,]+, 0`

> Storing every event inside a std::multimap and iterating over all its elements with every update tick has a bad performance.

EventMap doesnt seem to go through the entire multimap on each tick. It only iterates over the elements to execute. Infact, you use a multiset with same purpose as well.. so I guess this PR doesnt change that for the better, does it?

File: sql/custom/world/.gitignore (hunk 2 above)
Comment (Naios) [inline sql/custom/world/.gitignore]: Testing purposes like the other custom stuff as well.

**Comment by Naios** [comment]
@Lordron No, i wont do a performance comparison since you can't weight useability vs. performance, but probably the performance is not significantly faster or slower then EventMap.

@Rochet2 
> What does this mean? For example for events with 0 time, can they be executed instantly?

Depends on the context where you schedule the function in. If you schedule it inside a context its possible that its get executed immediatly (for example: a function is already 200ms behind and you schedule a function inside the other function in 100 ms it gets executed immediatly).
If you schedule it outside a context it wait for the next update tick.
I thought about add another overload vor schedule that accepts no time, such as a ScheduleAsync that will execute it at the next update tick.

> EventMap doesnt seem to go through the entire multimap on each tick.
Yeah you are right, was my mistake. I thought that multimap is unordered in every case. Probably its nearly the same as my implementation with std::multiset.

I don't want to say that EventMap is bad in general, but today it can be implemented much better using the suggested techniques.

**Comment by Naios** [comment]
/Push I need feedback on this to adapt it to the needs.
Do you like it or think it is unneccessary?
Is anything is missing for you, write it here then.

**Comment by Nyeriah** [comment]
Well, so instead of having the entire AI at one switch block we'll have it split in lambdas everywhere? That's sort of inconvenient from my point of view (yes the lambdas do look annoying)... but! That's just my opinion

**Comment by Naios** [comment]
Well, `std:.function` is only a wrapper for several types of functions which aren't limited to lambdas, we could also use:
* static methods or functions
* member functions
* lamdas (or lamdas prestored in variables).
* std::bind to prebind content to a function using placeholders.

I personally think lamdas look much better then a gigantic switch cases, but thats just my personal opinion.
But the code is much better readable if you define the eventbody where the event is scheduled, instead to have a huge UpdateAI which handles everything.

**Comment by Pesthuf** [comment]
It would be interesting to have hooks specifically for spell cast events. Currently with eventmaps, you often have the problem that the "cooldown" for a spell is applied even if the spell failed for some reason.

Like, when an NPC has a melee spell and the eventmap fires the event before the NPC is within Melee range of its target, the castSpell is called, the spell cast fails because he is not in range, but the spell "cooldown" is still reset in the AI(a new event is scheduled with no regard for whether the spell was successful or not).

On blizzard, the mob will immediately cast the spell when the spell can be cast (when he reaches you, in this example). It isn't just for melee spells, assuming a spell has a 30 yard cast range and a 3 second cast time, when the NPC starts the cast and its target runs out of the LOS, out of range or interrupts the cast by knockback or stun (but not by actual interrupt effect), the NPC will begin the cast again when it is back in LOS / in range, no longer stunned or knocked back. The "cooldown" seems to only be applied after the spell is successful (or kicked).

**Comment by Naios** [comment]
Yeah you are right, i thought about creating some async non-blocking callbacks like Node.js has it.
For example in Node you evaluate async database requests with a callback. Where you can check for errors, we could do the same with spellcasting.

```c++
me->CastSpell(me, 12345, [](bool success)
{
    if (!success)
        // reschedule
});
```
All in all the current scripting api is synchrone but most actions we invoke are async (spells, movement and dialogs), thats why we can't evaluate its result or react on it as soon as it is finished.
But i need more feedback on those ideas.

**Comment by DDuarte** [comment]
A.K.A callbacks... which eventually result in callback hell.

A task based approach would look a lot better.
https://casablanca.codeplex.com/wikipage?title=Programming%20with%20Tasks&referringTitle=Documentation

**Comment by Naios** [comment]
`std::future` is also a possible approach, by checking if the result is already ready in every update tick (bad solution in my opinion).

In general npc ai's are very **linear**, you wait for something, you yell something inplace, it resets and waits for the next enemy and so on... very similar to the good old **basic**, most actions occur step by step.

Thats why the following approach using a npc builder wouldn't be that bad:

```c++
enum Labels
{
    LBL_START
};

NpcBuilder::CreateEmptyBehaviour("npc_my_example_npc")
    .EnableReset()
    .EnableAutoAttack()
    .EnableSearchAndAttack()
    .WaitForEnterCombat()
    .Label(LBL_START)
        .Wait(std::chrono::seconds(20))
        .DisableSearchAndAttack()
        .Cast(37326)
        .Wait(std::chrono::seconds(7))
        .MoveTo({ 1.f, 1.f, 1.f, 1.f })
        .Cast(21927)
        .Inplace([](Creature* me)
        {
            // Do something special
            // me->
        })
        .Wait(std::chrono::seconds(4))
        .EnableSearchAndAttack()
        .GoTo(LBL_START)
    .Build();

```

We could generate an ai dynamically on top of the low level CreatureAI.

**Comment by Naios** [comment]
@DDuarte I took a deep look into casablanca, i like its fluent style forward logic (its very similar to some async js libraries).
The actual problem is that it works with threadpools where functions return its result when finished instead of callbacks (what we actually need), i also didn't figure out how to execute the `pplx::task` inplace on the mainthread (if it's possible).

**Comment by DDuarte** [comment]
https://github.com/TrinityCore/TrinityCore/commit/da77a90aa98bc6e34c001ca871700cec7c6e5017https://github.com/TrinityCore/TrinityCore/commit/da77a90aa98bc6e34c001ca871700cec7c6e5017

**Pull request closed without merging**